[DataGridPro] Filtering on Column Header

MBilalShafi opened the pull request:
Closes #6247 

- [x] Implement Header filtering base functionality
- [x] Synchronize width change
- [x] Add support to Pro
- [x] Add support to Premium
- [x] Introduce dedicated state for Header Filtering (state.headerFiltering)
- [x] Fix focus on menu close
- [x] Scope-out `isAnyOf` operator for this PR (Will be handled in: https://github.com/mui/mui-x/issues/9034)
- [x] Reorganize components
- [x] Handle cell events
- [x] Accessibility (ARIA attributes)
- [x] Keyboard Navigation
- [x] Prefix public methods with `unstable_`
- [x] Localization
- [x] Add a demo utilizing `colDef.renderHeaderFilter` and polish documentation
- [x] Testing

Nice to Haves (recommendations will be appreciated):
- [ ] Reexport from `/internals` (avoid eslint comment)
- [ ] Move headerFiltering state, keyboard navigation, useColumnResize, and focus logic to the Pro package (too tightly coupled, maybe leave it for another PR?)

Issue: [#6247](https://github.com/mui/mui-x/issues/6247)
Designs: https://www.figma.com/proto/W6TItYkiGJZjL5zLnfeSjy/Data-grid?node-id=1024%3A44648&scaling=scale-down-width&page-id=650%3A35241&starting-point-node-id=905%3A40731

## Preview
https://deploy-preview-7760--material-ui-x.netlify.app/x/react-data-grid/filtering/#header-filters

joserodolfofreitas commented on docs/data/data-grid/filtering/filtering.md:
```diff
@@ -47,6 +47,39 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
 
 {{"demo": "DisableActionButtonsDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
 
+## Header filtering [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)
```
I believe this sounds better, and other components also don't overuse the "ing" (we should probably consider dropping the "ing" from the other filtering titles too– on a different PR).

```suggestion
## Header filters [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)
```

MBilalShafi replied:
Adding another two cents: Maybe relate it to the quick filter but on a column level?
```suggestion
## Quick column filters [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)
```

Or simply `Column filter`, what do you think?

cherniavskii replied:
I think "Quick column filters" might be confusing given that we also have a "Quick filtering" feature.
"Header filters" sounds good to me though.

joserodolfofreitas commented on docs/data/data-grid/filtering/filtering.md:
````diff
@@ -47,6 +47,39 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
+This feature is experimental, it needs to be explicitly activated using the `headerFiltering` experimental feature flag.
+
+```tsx
+<DataGrid experimentalFeatures={{ headerFiltering: true }} {...otherProps} />
````
```suggestion
<DataGridPro experimentalFeatures={{ headerFiltering: true }} {...otherProps} />
```

joserodolfofreitas commented on docs/data/data-grid/filtering/filtering.md:
```diff
@@ -47,6 +47,39 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
+
+{{"demo": "HeaderFilteringDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
+
+:::info
```
I believe this `info` deserves its own section on the page with a live example.

joserodolfofreitas reviewed (commented):
Great work, @MBilalShafi! The solution looks and feels great!

I was wondering about some customization abilities that could be useful for developers:
Could it be a quick win to offer a `simplifiedFilterField` option where users can filter with only a default operator, but as a tradeoff, the grid doesn't display the fields' labels and operator controller?

Regarding keyboard navigation, I think it could be an improvement in usability if users can navigate to the next filter using tabs. Currently, users need two tabs to go to the next field. Is this feasible? What do you think about this?

DanailH commented on packages/grid/x-data-grid-pro/src/components/GridColumnHeaders.tsx:
```diff
@@ -253,6 +257,12 @@ const GridColumnHeaders = React.forwardRef<HTMLDivElement, DataGridProColumnHead
               },
               { disableReorder: true },
             )}
+
```
```suggestion
```

MBilalShafi replied:
`yarn proptypes` keeps inserting them again 😆

DanailH commented on packages/grid/x-data-grid-pro/src/components/GridColumnHeaders.tsx:
```diff
@@ -292,6 +307,12 @@ const GridColumnHeaders = React.forwardRef<HTMLDivElement, DataGridProColumnHead
               },
               { disableReorder: true, separatorSide: GridColumnHeaderSeparatorSides.Left },
             )}
+
```
```suggestion
```

DanailH commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterAdorment.tsx:
```diff
@@ -0,0 +1,70 @@
+          aria-expanded={open ? 'true' : undefined}
+          aria-haspopup="true"
+          tabIndex={-1}
+          size="small"
```
It's a broader question but should make the Adorment a slot (or maybe there is already a slot for the entire input including the adornment)?

MBilalShafi replied:
I made this and also menu a slot previously, as you said. Although it should be opinionated, but I decided not to go with it ultimately:
1. It's not just `Adorment` which the user has to reproduce, it's also the logic inside which manages state and encapsulates menu, so probably it's not that practical in real-world scenarios (the same problem that we have with the other filter panel)
2. It's easily achievable using anyways using `colDef.renderHeaderFilter` and also by `slotProps` (it's underway):
```tsx
<DataGridPro
  {...otherProps}
  slotProps={{
    headerFilter: {
      // props passed to root div in `<GridHeaderFilterItem />`,
      ...headerFilterItemPropsOverrides,
      // props passed to respective filter input component
      InputComponentProps: {
        InputProps: {
          startAdorment: <SomeCustomAdorment /> // but it should include menu inside to work like today
        }
      }
    }
  }}
/>
```

What do you think?

MBilalShafi replied:
> Can we move InputAdornment to filter input components, since it's now their implementation detail? (They receive headerFilterMenu prop which is then passed as startAdornment)

Haven't we already moved InputAdornment to the filter input components 🤔  (as `headerFilterMenu` encapsulates both `InputAdornment` and `GridHeaderFilterMenu`) or you mean to extract `InputAdornment` one level above and pass it as a separate prop like `headerFilterMenu`?

Wouldn't it make filter input components less generic and require more Menu related logic to them (as they are also being used a bit differently in filter panel)

Or am I not getting your point correctly 🤔

MBilalShafi replied:
> Also, adding a slot for it would be great

Added slots for Menu, Adornment, and Header filter cell

DanailH commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,279 @@
+          return;
+        }
+        if (isEditing) {
+          switch (event.key) {
```
Can we rework it to have a single `switch` and do the `isEditing` checks inside each `case`? I think it will make it easier to read. It's just a suggestion.

MBilalShafi replied:
Thanks for the suggestion, I tried to do the same earlier, but it seemed less manageable to me as compared to this, anyhow, I'll review it again and see if we can compactly reduce one level of indentation.

DanailH commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,279 @@
+              apiRef.current.stopHeaderFilterEditMode();
+            }}
+            fullWidth
+            placeholder={placeholder}
```
I think you can directly inline it.

```suggestion
            placeholder={apiRef.current.getLocaleText('columnMenuFilter')}
```

DanailH commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,279 @@
+            onFocus={() => {
+              apiRef.current.startHeaderFilterEditMode(colDef.field);
+            }}
+            onBlur={() => {
+              apiRef.current.stopHeaderFilterEditMode();
+            }}
```
```suggestion
            onBlur={apiRef.current.stopHeaderFilterEditMode}
```

DanailH commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,279 @@
+            item={item}
+            inputRef={inputRef}
+            applyValue={applyFilterChanges}
+            onFocus={() => {
+              apiRef.current.startHeaderFilterEditMode(colDef.field);
+            }}
```
```suggestion
            onFocus={() => apiRef.current.startHeaderFilterEditMode(colDef.field)}
```

DanailH commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,279 @@
+            }}
+            fullWidth
+            placeholder={placeholder}
+            label={isFilterActive ? OPERATOR_LABEL_MAPPING[item.operator] : ' '}
```
We need to think about making `OPERATOR_LABEL_MAPPING` localizable. I know that you haven't marked the localization as done but we can see if it woundn;t be too difficult to add it to the first version.

MBilalShafi replied:
That's a good point, I am actually thinking to update the labels in the original filter panel too, i.e wherever we have symbols instead of values, that way we can open them up for localization and make this consistent across filter panels.

Example of such labels:
https://github.com/mui/mui-x/blob/10162e2b973d93fd3b110e33ed83652cf5e3542b/packages/grid/x-data-grid/src/colDef/gridNumericOperators.ts#L30-L31

The `label` could be `equals` instead and then we can map it with the respective localizations.

MBilalShafi replied:
Added localization for all the operators except numeric, which I believe to be a breaking change as we have to change `operator.value` for those operators which have symbols used as value (=, !=, >, >=, < and <=)
We can cater this in v7 probably.

DanailH commented on packages/grid/x-data-grid-pro/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx:
```diff
@@ -0,0 +1,140 @@
+
+    const filters: JSX.Element[] = [];
+    for (let i = 0; i < renderedColumns.length; i += 1) {
+      const colDef = renderedColumns[i];
+
+      const columnIndex = firstColumnToRender + i;
+
+      const hasFocus = columnHeaderFilterFocus?.field === colDef.field;
+
+      const isFirstColumn = columnIndex === 0;
```
```suggestion
      const colDef = renderedColumns[i];
      const columnIndex = firstColumnToRender + i;
      const hasFocus = columnHeaderFilterFocus?.field === colDef.field;
      const isFirstColumn = columnIndex === 0;
```

DanailH reviewed (commented):
Awesome work @MBilalShafi 🚀. The feature is smooth and looks like a finished product. I left a couple of suggestions and questions.

One more thing - maybe makes sence to list this feature here: https://mui.com/x/react-data-grid/getting-started/#feature-comparison

MBilalShafi commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterAdorment.tsx:
```diff
@@ -0,0 +1,70 @@
+          aria-controls={menuId}
+          aria-expanded={open ? 'true' : undefined}
+          aria-haspopup="true"
+          tabIndex={-1}
```
Oops I forgot to document it, it's keyboard accessible the same way as a column menu button, i.e. by doing <kbd>Cmd</kbd> + <kbd>Enter</kbd> when the cell is focused. I'll add it.

MBilalShafi commented:
Thanks @joserodolfofreitas @DanailH for the useful feedback.

> Could it be a quick win to offer a `simplifiedFilterField` option where users can filter with only a default operator, but as a tradeoff, the grid doesn't display the fields' labels and operator dropdown?

It's a nice recommendation and will definitely smooth out some use-cases.

It should be possible for the developers to achieve a similar thing by props override (something like [this](https://github.com/mui/mui-x/pull/7760#discussion_r1165372091)), regarding supporting it natively via a prop though, I have a couple of points in my mind to ponder upon:
1. It might not work great with the synchronization with the main filter panel. I mean wouldn't it be confusing for the users to see a filter in the filter panel but not in the header filter (if that is not the default one). Some options to tackle it:
   - Isolate both the filters (header filter and regular filter panel) and their respective states, just like how the quick filter and regular filter panel are isolated. Probably also introduce a new entry in the `filterModel`, something like `filterModel.headerFilters`. (I would not be in favor of this as it will add a new layer of complexity)
   - Add a recipe that shows how to use `slotProps` to achieve a simple filter with no dropdown and placeholder.
2. Also, it's not very clear to me how to define a default operator in case of user defined filters like in [this demo](https://mui.com/x/react-data-grid/filtering/#custom-input-component). Options that we have:
   - Use `operators[0]` as the default filter operator
   - Extend API to allow user to configure the default operator for a column, something like: `getDefaultFilterOperator = (filterOperators: GridFilterOperator) => string`

What do you think?

> Regarding keyboard navigation, I think it could be an improvement in usability if users can navigate to the next filter using tabs. Currently, users need two tabs to go to the next field. Is this feasible? What do you think about this?

Thanks for pointing this out, the behavior on the live demo was actually an unintended behavior. I wanted it to work the same like a normal Grid cell. i.e.
- If some filter cell is in Edit mode, it's tabbable, pressing tab stops it from editing and focuses the very next cell (not making it editable though)
- When no filter cell is in edit mode, Tab normally, i.e. no filter cell is tabbable and users can go to the next natural tabbable element (let's say pagination)
- **Why not just like a row that is in edit mode (i.e. pressing tab moves focus to next editable cell):** Because I don't want to hijack the tabbable behavior of the grid. Imagine a person wants to go to next page by keyboard navigation only and the grid has 100 columns. They'll have to press Tab 101 times to focus on pagination component. Insane, right? 😄 

I updated the behavior like I mentioned, let me know your thoughts on this 🙏 

> One more thing - maybe makes sence to list this feature here: https://mui.com/x/react-data-grid/getting-started/#feature-comparison

Totally makes sense, thanks for the suggestion, I'll add it.

cherniavskii commented on packages/grid/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:
```diff
@@ -26,9 +30,36 @@ const DataGridPremiumRaw = React.forwardRef(function DataGridPremium<R extends G
 ) {
   const props = useDataGridPremiumProps(inProps);
   const privateApiRef = useDataGridPremiumComponent(props.apiRef, props);
+  const headerFiltersRef = React.useRef<HTMLDivElement>(null);
 
+  privateApiRef.current.register('private', {
+    headerFiltersElementRef: headerFiltersRef,
+  });
```
Maybe move it to `useGridColumnHeaders` or `GridBody` to avoid duplication in `DataGridPro` and `DataGridPremium`?

MBilalShafi replied:
Done ✅

joserodolfofreitas reviewed (commented):
Should we remove the filters from the model when they are emptied?
I believe it can be confusing if the user accesses the filtering panel after playing around with filters on header.

https://user-images.githubusercontent.com/550141/234805261-a0cccba4-b766-4036-8ab2-8e2955279919.mov

MBilalShafi commented on docs/data/data-grid/filtering/filtering.md:
```diff
@@ -47,6 +47,41 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
+{{"demo": "HeaderFilteringDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
+
+:::info
+**Tip:** You can render custom components in the header filtering row by using the `colDef.renderHeaderFilter` method.
```
Tried to do it in https://github.com/mui/mui-x/pull/7760/commits/54e0b17f51bc8d2e358acf964cd7d42a11c150f2 and [updated](https://codesandbox.io/s/3okcng?file=/demo.tsx) the codesandbox to reflect new changes, do you think it makes sense now?

cherniavskii replied:
> do you think passing the Adorment component with a generic name (let's say headerFilterMenu) and then using that prop inside default Filter components such as GridFilterInputValue etc to make it TextField agnostic and use in custom input components the same way would work?

Yes, I think we should not assume TextField API here.
 
> Tried to do it in https://github.com/mui/mui-x/commit/54e0b17f51bc8d2e358acf964cd7d42a11c150f2 and [updated](https://codesandbox.io/s/3okcng?file=/demo.tsx) the codesandbox to reflect new changes, do you think it makes sense now?

Looks awesome!
One thing I noticed here is that you have to click twice to open the menu if the cell is not focused:

https://user-images.githubusercontent.com/13808724/235452647-753fa66f-4f5c-42e9-96fe-f609185861cd.mov

cherniavskii replied:
> Another one would be to allow usecases like https://github.com/mui/mui-x/issues/4934#issuecomment-1240320900 (point no. 2) where users would like to use 2nd header row for some reason other than the filtering itself.

I'm unsure about using `renderHeaderFilter` for this, as it seems inefficient - a lot of filter-related computations will be done anyway.
But it will work.

MBilalShafi replied:
> a lot of filter-related computations will be done anyway.

Yes this concern is valid 👍 , one way of solving this that I thought of was to add a slot for the `HeaderFilterItem` and let the users override the whole slot instead of sub-part that is done by `renderHeaderFilter`, but I decided to not go with it for now due to these reasons.
1. It will require the users to receive props like width and height and apply properly for it to work as expected (learning curve)
2. It will be a bit more difficult to do customization.
    - A user want to disable header filter for a specific column.
      1. **Using `renderHeaderFilter`**, they can pass something like `renderHeaderFilter: () => null` in `colDef`
      2. **Using slot:** The user will have to conditionally pass `null` after checking `colDef.field === 'x'` and reproduce the default behavior themselves (copy paste all the default code) which doesn't seem great.
    - Another user want to use a custom header filter for a specific column.
      1. **Using `renderHeaderFilter`**, they can pass something like `renderHeaderFilter: (props) => <CustomHeaderFilter {...props} />` in `colDef`
      2. **Using slot:** The user will have to conditionally pass `<CustomHeaderFilter {...props} />` after checking `colDef.field === 'x'` and again reproduce the default behavior themselves.
    - What if they want to override/customize for all columns?
      Should we introduce something like `slotProps.headerFilter.renderHeaderFilter` which is preceded for the columns where `colDef.renderHeaderFilter` is not passed 🤔 
4. So going with the same pattern like `renderHeader`, `renderCell` and `renderHeaderGroup` seemed more like it.

We can consider point no. 2 mentioned [here](https://github.com/mui/mui-x/pull/7760#discussion_r1179251451) as a corner case which is handled as a bonus along with the basic functionality.

Let me know if you disagree on something as this is a open to discussion topic.

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,298 @@
+        aria-colindex={colIndex + 1}
+        aria-label={headerFilterComponent == null ? colDef.headerName ?? colDef.field : undefined}
+        {...other}
+        {...mouseEventsHandlers}
```
I created this sandbox with one column that is not filterable:
https://codesandbox.io/s/floral-wildflower-8obfkj?file=/demo.tsx

And after clicking on the empty header filter, the keyboard navigation doesn't work anymore (it enters editing mode)

https://user-images.githubusercontent.com/13808724/235479991-789e6703-b319-4f1b-87c1-89b91fb425a1.mov

Maybe we can return earlier and avoid that?

Nit: Another option would be offloading keyboard navigation to header filter components.
If we allow them to control whether navigation keys should move the focus to a different cell or not, then we might not even need the global state for this.

cherniavskii replied:
Same happens when using `renderHeaderFilter`: https://codesandbox.io/s/delicate-thunder-yxjmcv?file=/demo.tsx

MBilalShafi replied:
Thanks for reporting the issue, the reason why it was happening like this is because we were deliberately moving the cell to edit mode on cell click, which doesn't make sense especially when the filter cell is empty.

I [tried](https://github.com/mui/mui-x/pull/7760/commits/45e6be0d5d8d8ac50bfcc6f2292c58e3b72cd919) to fix it, here's the updated codesandbox: https://codesandbox.io/s/aged-voice-jnvow4?file=/demo.tsx

Let me know your thoughts on it?

> Nit: Another option would be offloading keyboard navigation to header filter components.
If we allow them to control whether navigation keys should move the focus to a different cell or not, then we might not even need the global state for this.

You mean move remove the Header Filters' related logic from `useKeyboardNavigation` and `useGridFocus` and move focus and keyboard logic to GridHeaderFilterItem component?

cherniavskii replied:
> You mean move remove the Header Filters' related logic from useKeyboardNavigation and useGridFocus and move it to GridHeaderFilterItem component?

I was thinking about being able to enable/disable keyboard navigation by the cell (e.g. when in edit mode).
But I don't think it's the way to go.
For the custom header filter components, they can use `startHeaderFilterEditMode`/`stopHeaderFilterEditMode` to leverage the editing state, right? If so, then we should be good 👍

cherniavskii commented on packages/grid/x-data-grid/src/hooks/features/headerFiltering/useGridHeaderFiltering.ts:
```diff
@@ -0,0 +1,120 @@
+    (headerFilterState: Partial<GridHeaderFilteringState>) => {
+      apiRef.current.setState((state) => {
+        // Safety check to avoid MIT users from using it
+        // This hook should ultimately be moved to the Pro package
```
Is the challenge here that these methods are being used by the code in the MIT package?

MBilalShafi replied:
Yes, we have to keep the state (and related methods) in MIT due to tight coupling in hooks like `useGridFocus`, `useGridKeyboardNavigation`

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx:
```diff
@@ -0,0 +1,298 @@
+        {...mouseEventsHandlers}
+      >
+        {headerFilterComponent}
+        {InputComponent && !headerFilterComponent ? (
```
It looks like `renderHeaderFilter: () => null` doesn't work as expected and renders the default component:
https://codesandbox.io/s/boring-colden-m260of?file=/demo.tsx

MBilalShafi replied:
Fixed, it should be working now
https://codesandbox.io/s/aged-voice-jnvow4?file=/demo.tsx

MBilalShafi commented on packages/grid/x-data-grid/src/hooks/features/headerFiltering/useGridHeaderFiltering.ts:
```diff
@@ -0,0 +1,120 @@
+  const showHeaderFilterMenu = React.useCallback<GridHeaderFilteringApi['showHeaderFilterMenu']>(
+    (field) => {
+      logger.debug(`Opening header filter menu for field: ${field}`);
+      apiRef.current.setHeaderFilterState({ menuOpen: field });
```
I think because we need to check for it in useGridKeyboardNavigation
https://github.com/MBilalShafi/mui-x/blob/5002c7aebb3be0c8bce516a354ac88847588628d/packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts#L317-L318

cherniavskii replied:
No, I was just wondering if it's indeed necessary. I didn't notice that we need it for keyboard navigation 👍

romgrk commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/constants.ts:
```diff
@@ -0,0 +1,52 @@
+  '<=': 'is less than or equal to',
+};
+
+export const OPERATOR_SYMBOL_MAPPING: { [key: string]: string } = {
```
In these cases, wouldn't it be better to let tsc infer the typing? The infered type would be more strict than just `{ [key: string]: string }`.

MBilalShafi replied:
Thanks for the suggestion, I tried but strict type here doesn't play well with the [type](https://github.com/mui/mui-x/blob/1f0f06e4697a991388d06d4bdf16aa9da4800780/packages/grid/x-data-grid/src/models/gridFilterItem.ts#L24) of `GridFilterItem.operator` which we have to use as a key for this object. Because that way we may have to typecast `GridFilterItem.operator` to an incorrect type (i.e. the strict type inferred by `tsc`) to make `tsc` happy.

What do you think?

MBilalShafi commented on packages/grid/x-data-grid-pro/src/DataGridPro/DataGridPro.tsx:
```diff
@@ -28,6 +32,29 @@ const DataGridProRaw = React.forwardRef(function DataGridPro<R extends GridValid
+    columnHeaderFilterTabIndexState === null &&
+    columnHeaderTabIndexState === null &&
+    cellTabIndexState === null
+  );
```
Moved to the `useGridColumnHeaders` hook on the Pro package.

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/GridColumnHeaders.tsx:
```diff
@@ -267,6 +274,11 @@ const GridColumnHeaders = React.forwardRef<HTMLDivElement, DataGridProColumnHead
             minFirstColumn: leftPinnedColumns.length,
             maxLastColumn: visibleColumnFields.length - rightPinnedColumns.length,
           })}
+          {getColumnFilters({
```
There's a slight layout jump when navigating between filter row and column headers:

https://github.com/mui/mui-x/assets/13808724/d734d705-8b19-4968-b465-b2648e69efda

MBilalShafi replied:
Interesting. It's not reproducible on Firefox but on Chrome, and the same is also observed with [column grouping](https://mui.com/x/react-data-grid/column-groups/#column-reordering).

![groupedColumns](https://github.com/mui/mui-x/assets/12609561/cfde6cab-7a43-4247-af0a-50bcc412c9c9)

Seems like a CSS thing, will investigate more. Thanks for reporting.

cherniavskii commented:
> Added an opt-in clear icon on header filters as discussed: [deploy-preview-7760--material-ui-x.netlify.app/x/react-data-grid/filtering/#clear-filter-icon](https://deploy-preview-7760--material-ui-x.netlify.app/x/react-data-grid/filtering/#clear-filter-icon)
Does it make sense to make it enabled by default?

IMO enabling it by default totally makes sense 👍

cherniavskii commented on docs/data/data-grid/filtering/filtering.md:
```diff
@@ -47,6 +47,80 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
+
+#### Customize using `filterOperators`
+
+You can also update the `InputComponent` of the `filterOperator` using `GridColDef.filterOperators` to customize the header filter cell for a specific operator type.
```
Maybe refer to https://mui.com/x/react-data-grid/filtering/#custom-input-component and remove the `disableColumnFilter` from the demo to make it clear that the same custom component is being used for both header filter and filter panel? Something like this:

```suggestion
If the filter operator has a [custom `InputComponent`](https://mui.com/x/react-data-grid/filtering/#custom-input-component), the same component is being used for the header filter.
When rendered as a header filter, the `InputComponent` also receives the `headerFilterMenu` prop that contains the filter operator menu. 
```

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:
```diff
@@ -0,0 +1,355 @@
+                buttonRef={buttonRef}
+              />
+            }
+            clearIcon={
+              showClearIcon ? (
+                <rootProps.slots.headerFilterClearIcon onClick={clearFilterItem} />
+              ) : null
+            }
```
Then in a custom `InputComponent` we can do:

```tsx
{headerFilterMenu}
{/** **/}
{clearIcon}
```

cherniavskii replied:
Also, rename it to `clearButton`?

MBilalShafi replied:
> Maybe pass it conditionally, only when isApplied === true?

I think it makes sense, I didn't do it previously as I wanted to give more control to user when to render the icon, but it's pointless to render it anyways when no filter is applied.

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterAdornment.tsx:
```diff
@@ -0,0 +1,107 @@
+  return (
+    <React.Fragment>
+      <InputAdornment position="start">
+        <rootProps.slots.baseIconButton
```
Maybe make it a proper circle? Currently, it has an elliptic shape:


https://github.com/mui/mui-x/assets/13808724/b7b37735-f477-4711-a698-fc4180df3187

cherniavskii replied:
Looks better now!
Although some symbols are not centered: 

<img width="135" src="https://github.com/mui/mui-x/assets/13808724/cbee614b-2314-4b24-9535-a98d11ff0f3f">

MBilalShafi commented on packages/grid/x-data-grid-pro/src/constants/dataGridProDefaultSlotsComponents.ts:
```diff
@@ -2,11 +2,19 @@ import { DATA_GRID_DEFAULT_SLOTS_COMPONENTS } from '@mui/x-data-grid/internals';
+  HeaderFilterAdornment: GridHeaderFilterAdornment,
+  HeaderFilterCell: GridHeaderFilterCell,
+  HeaderFilterMenu: GridHeaderFilterMenu,
+  HeaderFilterClearIcon: GridHeaderFilterClearIcon,
```
Correct 👍

MBilalShafi commented on packages/grid/x-data-grid-pro/src/constants/dataGridProDefaultSlotsComponents.ts:
```diff
@@ -2,11 +2,19 @@ import { DATA_GRID_DEFAULT_SLOTS_COMPONENTS } from '@mui/x-data-grid/internals';
   ...materialSlots,
   ColumnMenu: GridProColumnMenu,
   ColumnHeaders: GridColumnHeaders,
+  HeaderFilterAdornment: GridHeaderFilterAdornment,
```
Yes, it makes more sense 👍

MBilalShafi replied:
Done ✅

cherniavskii commented on docs/data/data-grid/filtering/CustomHeaderFilterOperatorDataGridPro.js:
```diff
@@ -0,0 +1,121 @@
+          display: 'inline-flex',
+          flexDirection: 'row',
+          alignItems: 'center',
+          height: 20,
```
```suggestion
          height: '100%',
```
To fix alignment in the filter panel

cherniavskii commented on docs/data/data-grid/filtering/CustomHeaderFilterOperatorDataGridPro.js:
```diff
@@ -0,0 +1,121 @@
+          precision={0.5}
+          ref={ratingRef}
+        />
+      </Box>
```
Add `{clearButton}` here?

MBilalShafi commented on scripts/x-data-grid-premium.exports.json:
```diff
@@ -49,6 +49,7 @@
   { "name": "getGridBooleanOperators", "kind": "Variable" },
   { "name": "getGridDateOperators", "kind": "Variable" },
   { "name": "getGridDefaultColumnTypes", "kind": "Variable" },
+  { "name": "getGridFilter", "kind": "Variable" },
```
Yes, I guess it was needed in `useGridColumnHeaders` (pro version)

MBilalShafi commented on packages/grid/x-data-grid-pro/src/models/gridProSlotProps.ts:
```diff
@@ -0,0 +1,6 @@
+import type { GridHeaderFilterCellOverridableProps } from '../components/headerFiltering/GridHeaderFilterCell';
+
+export interface GridProSlotProps extends GridSlotsComponentsProps {
+  headerFilter?: Partial<GridHeaderFilterCellOverridableProps>;
```
Yes, see in this demo for example: https://deploy-preview-7760--material-ui-x.netlify.app/x/react-data-grid/filtering/#simple-header-filters

cherniavskii replied:
Right, I didn't notice that 👍
Should it be renamed to `headerFilterCell` to match the slot name?

Also, it seems to me that the most straightforward way of removing the header filter menu would be:
```tsx
slots={{
  headerFilterMenu: null,
}}
```
Instead of 
```tsx
slotProps={{
  headerFilter: {
    InputComponentProps: {
      headerFilterMenu: null,
    },
  },
}}
```

MBilalShafi replied:
Agreed, though it could still be useful for overriding other props like `InputComponentProps` etc

MBilalShafi commented on docs/data/data-grid/filtering/filtering.md:
````diff
@@ -47,6 +47,74 @@ To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` o
+This feature is experimental, it needs to be explicitly activated using the `headerFilters` experimental feature flag.
+
+```tsx
+<DataGridPro experimentalFeatures={{ headerFilters: true }} {...otherProps} />
````
I guess so 👍

MBilalShafi commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterAdornment.tsx:
```diff
@@ -0,0 +1,111 @@
+  const buttonId = useId();
+  const menuId = useId();
+
+  const rootProps = useGridRootProps() as DataGridProProcessedProps;
```
Nice, I didn't know it existed 😄

MBilalShafi commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/constants.ts:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Column types that don't need to render a filter cell.
+ */
+export const TYPES_WITH_NO_FILTER_CELL = ['actions', 'checkboxSelection'];
```
Indeed 🤦‍♂️

MBilalShafi commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/constants.ts:
```diff
@@ -0,0 +1,53 @@
+export const OPERATOR_LABEL_MAPPING: { [key: string]: string } = {
```
That's a good idea, I actually thought to extend the existing way of handling these labels, i.e. convert `value` for the operators (which are symbols) to strings (all in [gridNumericOperators.ts](https://github.com/mui/mui-x/blob/master/packages/grid/x-data-grid/src/colDef/gridNumericOperators.ts)), then we can define these literals in `localeTextConstants.ts` below these:
https://github.com/mui/mui-x/blob/0f70a86ba1816e9a3c2dc6383c275bc49d964fec/packages/grid/x-data-grid/src/constants/localeTextConstants.ts#L59-L71

And use them like this (as being done in GridFilterForm):
https://github.com/mui/mui-x/blob/0f70a86ba1816e9a3c2dc6383c275bc49d964fec/packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterForm.tsx#L517-L520

That way we can:
1. Centralize labels for both regular filter form and header filters
2. Allow all the labels to be translated

But since this seems a breaking change, I [decided](https://github.com/mui/mui-x/pull/7760#discussion_r1171326330) to postpone it till v7 and use constants for now.

Do you think this a viable approach or do you have more use-cases of `headerFilterLabel` in mind which are not possible with the approach I mentioned?

cherniavskii replied:
Regarding numeric operators - I think there's nothing blocking us from adding them to `localeTextConstants`:
```diff
filterOperatorIsAnyOf: 'is any of',
+'filterOperator=': '=',
```
And then removing the `label` in `gridNumericOperators`. I don't see this as a breaking change, unless I'm missing something

For the header filter labels, we can use the same approach - add `headerLabel`, and then add `headerFilterOperator${operator}` to `localeTextConstants`.
The `headerLabel` is optional (same as `label`), and if it's not defined - we call:

```tsx
apiRef.current.getLocaleText(`headerFilterOperator${capitalize(item.operator!)}
```

Does it make sense?

MBilalShafi replied:
> Regarding numeric operators - I think there's nothing blocking us from adding them to localeTextConstants:

Correct, I just wanted them to look a bit cleaner, so instead of `filterOperator=` => `filterOperatorEquals` which would require updating `operator.value` for the respective operators.

I was also staying away from introducing new locale text constants for the operators whose values would be exactly the same as `filterOperator${operator}` (doesn't make a lot of sense to have duplicate values, unless they could become different for some use-cases).

But, I agree that doing so would allow more customization by possibly having different labels for header filters and normal filters and also to the developers who want to define/customize custom filter operators.

MBilalShafi replied:
Done ✅ in https://github.com/mui/mui-x/pull/7760/commits/fd2264418a0107bb2f5cc197661a7fbc1389b0c0 and https://github.com/mui/mui-x/pull/7760/commits/49086949274396cd42940a65e15bbb725eec5d37

MBilalShafi commented on packages/grid/x-data-grid/src/models/events/gridEventLookup.ts:
```diff
@@ -192,6 +192,37 @@ export interface GridColumnHeaderEventLookup {
   };
 }
 
+export interface GridHeaderFilterEventLookup {
```
Sure thing 👍

MBilalShafi commented on scripts/x-data-grid-premium.exports.json:
```diff
@@ -107,6 +108,7 @@
   { "name": "GridArrowUpwardIcon", "kind": "Variable" },
   { "name": "GridAutoGeneratedGroupNode", "kind": "Interface" },
   { "name": "GridAutoGeneratedPinnedRowNode", "kind": "Interface" },
+  { "name": "GridBaseSlots", "kind": "Interface" },
```
I introduced this to type `materialSlots` which was not checking for incorrect usages previously.
https://github.com/mui/mui-x/blob/0f70a86ba1816e9a3c2dc6383c275bc49d964fec/packages/grid/x-data-grid/src/material/index.ts#LL80C3-L80C3

Do you think it's useful to keep it in exports or should we remove it?

cherniavskii replied:
I would avoid increasing the API surface if possible to be less limited when changing internals 🙂

MBilalShafi replied:
Agreed, will remove the export!

MBilalShafi commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:
```diff
@@ -0,0 +1,367 @@
+import { GridHeaderFilterClearButton } from './GridHeaderFilterClearButton';
+import { OPERATOR_LABEL_MAPPING, NO_INPUT_OPERATORS, TYPES_WITH_NO_FILTER_CELL } from './constants';
+
+type GridHeaderFilterCellConditionalProps =
```
[Added](https://github.com/mui/mui-x/pull/7760/commits/6cf877b9e46208cc997346b0afb066756bc5ce56) a generic type for user-defined operators.

cherniavskii replied:
But is the `GridHeaderFilterCellConditionalProps` union even necessary? Shouldn't something like this be enough?
```tsx
interface GridHeaderFilterCellConditionalProps {
  colType: GridColDef['type'];
  operator: GridFilterOperator['value'];
  InputComponentProps: GridFilterOperator['InputComponentProps'];
}
```

By the way, the `operator` prop seems to be unused, so `colType` and `InputComponentProps` should be enough here.

cherniavskii replied:
And it probably could be moved to `GridHeaderFilterCellProps` interface

cherniavskii replied:
`colType` seems to be unused either

oliviertassinari commented:
I had a quick look at https://deploy-preview-7760--material-ui-x.netlify.app/x/react-data-grid/filtering/#header-filters, I'm coming from https://www.notion.so/mui-org/Drew-McManus-Upstage-CRM-53e2f83bd0ab46e59279835762d6777e. Things related to design that I noticed:

1. The design of the loading indicator feels off. I would expect a different color (lighter) & different size (smaller) to match the other clear icon:

<img width="185" alt="Screenshot 2023-05-12 at 01 40 03" src="https://github.com/mui/mui-x/assets/3165635/f9bcd91c-bc4e-4b34-abab-604f9ad7b712">

2. I was confused by the cell focus state that changes the text field display. I would have expected the textbox focus state to trigger this behavior rather than the cell focus. Not sure 100%, maybe it's more about point 3. that makes it feel off:

<img width="197" alt="Screenshot 2023-05-12 at 01 42 12" src="https://github.com/mui/mui-x/assets/3165635/d2bd1b85-1582-440f-9d23-150c2e49d8c9">

3. The cell focus state triggers feel wrong to me. Shouldn't it trigger on the cell focus event? Why does it wait for the mouse-up event? It's linked to point 2, it's inconsistent, fixing 2 or 3 would solve this.

https://github.com/mui/mui-x/assets/3165635/6d6f59e3-c5b7-49d7-b429-7f58560013d2

I can't notice anything else, niiice.

MBilalShafi commented:
Thanks @oliviertassinari for the useful feedback.

1. Agreed, updated the color and size to match the clear icon. (It is also updated on the regular filter panel and it feels better there too)
2. I think that's a fair point too, for cells with no value, on one side it feels like a distracting visual movement, to show and hide the label and menu icon on keyboard navigation, on the other, removing it may result in less visual feedback about the default filter operator. I've updated it to as you suggested but I'd be interested to know what the team thinks about it.
3. Yes, it's something broader than the scope of this PR, we plan to fix it separately. (Context: https://github.com/mui/mui-x/pull/7760#discussion_r1188514018)
4. Interesting observation, I've updated it to listen to the mouse down instead on click, which seems to solve the mentioned issues.

oliviertassinari commented:
1. It looks better 👍. To consider how the fix will work with Joy UI
2. The behavior looks correct 👍
3. Right, we could open a GitHub issue, I can't reproduce it on HEAD though.
4. The behavior looks correct 👍. I'm not sure about onMouseDown though, I feel like onFocus & onBlur would be closer to the state `isFilterActive` is meant to hold. It seems that we already do it:

```jsx
            onFocus={() => apiRef.current.startHeaderFilterEditMode(colDef.field)}
            onBlur={() => apiRef.current.stopHeaderFilterEditMode()}
```

so maybe we don't need `onMouseDown` (a speculation, I didn't try)

6. Something new that I noticed, when pressing <kbd>Enter</kbd> with the filter input focused, it blurs it. I believe we don't want this behavior since the value is updated in real-time.

MBilalShafi commented:
2. This one has a downside in its current behavior, since we are not showing the menu button on focus, but only on the editing state, the user cannot change the operator with the mouse or keyboard (<kbd>Cmd</kbd> + <kbd>Enter</kbd>) until they enter some value, as editing mode is stopped on blur, so it might not be a great UX.
![interaction](https://github.com/mui/mui-x/assets/12609561/f3522c91-9c8c-483a-b7a4-62fdd1b8a806)

What do you think?

5. You are right, <kbd>Enter</kbd> here doesn't serve to update the value as the update is done in a denounced way. The primary rationale for blurring the input field on <kbd>Enter</kbd> is to make keyboard navigation work as expected, because, during the editing mode, we pass the arrow keys down to the input field to let the user maneuver in the text freely, on blurring (i.e. when enter is pressed), the editing mode is stopped, and the arrow keys again work in cell navigation mode.


> I'm not sure about onMouseDown though, I feel like onFocus & onBlur would be closer to the state isFilterActive is meant to hold. It seems that we already do it:

I think an event here (`onMouseDown` or `onClick` etc) would be necessary because we also have to set the focus state (`apiRef.current.setColumnHeaderFilterFocus`) to the current field which in turn sets the correct focus to the respective filter cell and input. It may partially work with `onFocus`, bus since the event is binded to the input component rather than the cell, it will not work when the user clicks on the white area (around the text field)

romgrk commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterAdornment.tsx:
```diff
@@ -0,0 +1,115 @@
+          tabIndex={-1}
+          size="small"
+          onClick={handleClick}
+          sx={{
+            width: 22,
+            height: 22,
+            padding: '0px 0px 2px 2px',
+          }}
```
The `sx` value could be extracted, it would avoid continuously re-allocating it.

MBilalShafi replied:
Nice recommendation, applied 👍

cherniavskii commented on packages/grid/x-data-grid-pro/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx:
```diff
@@ -0,0 +1,155 @@
+          headerClassName={headerClassName}
+          filterOperators={filterOperators}
+          data-field={colDef.field}
+          colType={colDef.type as any}
+          operator={item.operator as any}
```
```suggestion
          colType={colDef.type}
          operator={item.operator}
```

cherniavskii commented on packages/grid/x-data-grid-pro/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx:
```diff
@@ -0,0 +1,155 @@
+          ? colDef.headerClassName({ field: colDef.field, colDef })
+          : colDef.headerClassName;
+
+      // TODO: Support for `isAnyOf` operator
```
This one is challenging because you can select multiple values here, but the header filter row height is fixed and cannot grow as opposed to the filter panel:

<img width="713" src="https://github.com/mui/mui-x/assets/13808724/edd5ea97-2edc-4b7a-947b-6e624d00c9a0">

https://mui.com/x/react-data-grid/filtering/

But we can display the number of selected values instead of individual tags, something like this: https://codesandbox.io/s/headless-pond-h0dwn9?file=/demo.tsx

We can do this in a follow-up PR though 🙂

cc @gerdadesign

MBilalShafi replied:
Good idea, we can iterate over the proposed design and do a follow-up. Thanks 👍

DanailH replied:
Maybe it's a good idea to create a ticket with all the follow-ups so we don't loose track of them (there isn't one already).

MBilalShafi replied:
Yeah, that's a good idea, I'll create one!

DanailH commented:
@MBilalShafi I checked it one more time - looks solid, and nicely done. 
There were a couple of things that were pointed out that can be addressed later - probably the most annoying one is the small jump when you focus and unfocus a header filter cell. I propose we create an issue and list all those things so we don't lose track of them.
Besides that, I'm ok with releasing the first version to see what the feedback will be (should be safe as it's marked as `unstable`).

cherniavskii reviewed (commented):
Great work, @MBilalShafi! 🎉

MBilalShafi commented:
Thanks everyone who contributed for the great review feedback to shape this PR up 🙏

oliviertassinari commented:
> The primary rationale for blurring the input field on Enter is to make keyboard navigation work as expected

Why would end users expect <kbd>Enter</kbd> to blur an input? This is an input inside an header. The notion of header cell sounds more like an implementation detail.

oliviertassinari commented on packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:
```diff
@@ -0,0 +1,321 @@
+            onFocus={() => apiRef.current.startHeaderFilterEditMode(colDef.field)}
+            onBlur={() => apiRef.current.stopHeaderFilterEditMode()}
+            placeholder={apiRef.current.getLocaleText('columnMenuFilter')}
+            label={isFilterActive ? capitalize(label) : ' '}
```
I think that we should change this logic, turning the label to an empty string breaks the floating label animation.

MBilalShafi replied:
Agreed, one reason it happens is that label also changes on focus (etc `Filter` -> `Contains`), added to the list of follow-up [items](https://github.com/mui/mui-x/issues/9034).
